### PR TITLE
Key-value execution tags

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3960

## Why are the changes needed?
The current execution tag is a hash tag (key-only tag), but we want to support a key-value tag. The key is required, but the value is optional. That said, you can still use a hash tag by not providing a value.

### Tables
- Drop `execution_admin_tags` table
- Drop `admin_tags` table
- Add `execution_tags` table

To keep backward compatibility, both labels and tags in the executionSpc will be added to the new `execution_tags` table.

Endpoint to get the workflows with tags
- Get: http://localhost:30080/api/v1/executions/flytesnacks/development/f6e2ba11a96c24e79a88
- List: http://localhost:30080/api/v1/executions/flytesnacks/development?limit=5
- List with filter: http://localhost:30080/api/v1/executions/flytesnacks/development?filters=value_in(execution_tag.key,flyte%3Btest)&limit=25&sort_by.direction=DESCENDING&sort_by.key=created_at
- List with filter: http://localhost:30080/api/v1/executions/flytesnacks/development?filters=value_in(execution_tag.value,hello%3Brust)&limit=25&sort_by.direction=DESCENDING&sort_by.key=created_at

The old filter still works, and it will get the tag from the new table
- List with filter: http://localhost:30080/api/v1/executions/flytesnacks/development?filters=value_in(admin_tag.name,flyte%3Btest)&limit=25&sort_by.direction=DESCENDING&sort_by.key=created_at

## How was this patch tested?
local sandbox

```bash
pyflyte run --remote --label tokio=3 --label rust=6 flyte-example/getting_started.py wf
pyflyte run --remote --tag 123 --tag 456 flyte-example/getting_started.py wf 
```

### Screenshots
Screenshot 2024-06-06 at 3 50 34 PM

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

File: flyteadmin/pkg/repositories/config/migrations.go [hunk 1]
```diff
@@ -1182,6 +1182,45 @@ var NoopMigrations = []*gormigrate.Migration{
...
+			return tx.Migrator().DropTable("execution_admin_tags")
+		},
+	},
+
+	{
+		ID: "2024-06-06-drop-admin-tags",
+		Migrate: func(tx *gorm.DB) error {
+			return tx.Migrator().DropTable("admin_tags")
+		},
+	},
```
Comment (eapolinario) [inline flyteadmin/pkg/repositories/config/migrations.go:1223]: Move these migrations to `ContinuedMigrations` on line 1231.

**Review by eapolinario — COMMENTED** [review]
This is looking pretty good. Left a few comments.

File: flyteadmin/pkg/repositories/models/execution.go [hunk 2]
```diff
@@ -72,6 +72,15 @@ type AdminTag struct {
 	Name string `gorm:"index:,unique;size:255"`
 }
 
+type ExecutionTag struct {
+	BaseModel
+	ExecutionKey
+	// The key of the tag.
+	Key string `gorm:"primary_key" valid:"length(0|255)"`
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flyteadmin/pkg/repositories/models/execution.go:79]: https://gorm.io/docs/models.html#Fields-Tags
it looks like Gorm doesn't support length, but we use it everywhere...

File: flyteadmin/pkg/repositories/config/migrations.go (hunk 1 above)
Comment (pingsutw) [reply flyteadmin/pkg/repositories/config/migrations.go]: done

File: flyteadmin/pkg/repositories/models/execution.go [hunk 3]
```diff
@@ -72,6 +72,15 @@ type AdminTag struct {
...
+	// The value of the tag.
+	Value string `gorm:"primary_key" valid:"length(0|255)"`
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flyteadmin/pkg/repositories/models/execution.go]: make sense, thanks

File: flyteadmin/pkg/repositories/config/migrations.go [hunk 4]
```diff
@@ -1182,6 +1182,45 @@ var NoopMigrations = []*gormigrate.Migration{
 			return tx.AutoMigrate(&Execution{})
 		},
 	},
+
+	{
+		ID: "2024-06-06-execution-tags",
+		Migrate: func(tx *gorm.DB) error {
+			return tx.Transaction(func(_ *gorm.DB) error {
+				// Create an execution_tags Table
+				if err := tx.AutoMigrate(&models.ExecutionTag{}); err != nil {
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flyteadmin/pkg/repositories/config/migrations.go]: done

File: flyteadmin/pkg/repositories/models/execution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (eapolinario) [reply flyteadmin/pkg/repositories/models/execution.go]: yeah, gorm only consumes the `gorm` tags. We might have had plans to use [govalidator](https://github.com/asaskevich/govalidator), but I can't see a single mention of actual usage in the repo history (i.e. run `git log -p -S '"github.com/asaskevich/govalidator"'`).

**Review by eapolinario — APPROVED** [review]

**Pull request merged**